crates/lgn-graphics-renderer/src/core/services/cameras.rs

```diff
@@ -1,12 +1,205 @@
+    let up = final_transform.up();
+    let right = final_transform.right();
+
+    let half_v_side = z_far * (fov_y.radians() * 0.5).tan();
+    let half_h_side = half_v_side * aspect_ratio;
+
+    let near_face_point = eye + forward * z_near;
+    let near_normal = -forward;
+    let near_plane: Float4 = Vec4::from((near_normal, -near_normal.dot(near_face_point))).into();
+
+    let far_face_point = eye + forward * z_far;
+    let far_normal = forward;
+    let far_plane: Float4 = Vec4::from((far_normal, -far_normal.dot(far_face_point))).into();
+
+    let front_mult_far = z_far * forward;
+
+    let right_side = front_mult_far - right * half_h_side;
+    let right_normal = up.cross(right_side).normalize();
+    let right_plane: Float4 = Vec4::from((right_normal, -right_normal.dot(eye))).into();
+
+    let left_side = front_mult_far + right * half_h_side;
+    let left_normal = left_side.cross(up).normalize();
+    let left_plane: Float4 = Vec4::from((left_normal, -left_normal.dot(eye))).into();
+
+    let top_side = front_mult_far - up * half_v_side;
+    let top_normal = top_side.cross(right).normalize();
+    let top_plane: Float4 = Vec4::from((top_normal, -top_normal.dot(eye))).into();
+
+    let bottom_side = front_mult_far + up * half_v_side;
+    let bottom_normal = right.cross(bottom_side).normalize();
+    let bottom_plane: Float4 = Vec4::from((bottom_normal, -bottom_normal.dot(eye))).into();
+
+    [
+        near_plane,
+        far_plane,
+        right_plane,
+        left_plane,
+        top_plane,
+        bottom_plane,
+    ]
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct RenderCamera<H: Handedness = RightHanded> {
+    pub final_transform: Transform<H>,
+    fov_y: Angle,
+    z_near: f32,
+    z_far: f32,
+}
 
 impl From<(&GlobalTransform, &CameraComponent)> for RenderCamera {
-    fn from(_: (&GlobalTransform, &CameraComponent)) -> Self {
-        RenderCamera {}
+    fn from((_global_transform, camera_component): (&GlobalTransform, &CameraComponent)) -> Self {
+        RenderCamera::new(camera_component) // TMP
+    }
+}
+
+impl RenderCamera {
+    pub fn new(camera_component: &CameraComponent) -> Self {
+        Self {
+            final_transform: camera_component.final_transform(),
+            fov_y: camera_component.fov_y(),
+            z_near: camera_component.z_near(),
+            z_far: camera_component.z_far(),
+        }
+    }
+
+    pub fn view_transform(&self) -> GlobalTransform {
+        view_transform(&self.final_transform)
+    }
+
+    pub fn build_projection(&self, width: f32, height: f32) -> Mat4 {
+        build_projection(width, height, self.fov_y, self.z_near)
+    }
+
+    pub fn build_culling_planes(&self, aspect_ratio: f32) -> [Float4; 6] {
+        build_culling_planes(
+            &self.final_transform,
+            aspect_ratio,
+            self.fov_y,
+            self.z_near,
+            self.z_far,
+        )
+    }
+}
+
+pub struct RenderCameraPrivateData {}
```

Comment:
Do you need this?

Reply:
The private data, not really. I'll remove it.